epoxy-adapter/src/main/java/com/airbnb/epoxy/AsyncListDifferWithPayload.java

```diff
@@ -0,0 +1,193 @@
+  private final static Executor MAIN_THREAD_EXECUTOR = new MainThreadExecutor();
+  private final ListUpdateCallback updateCallback;
+  private final ItemCallback<T> diffCallback;
+
+  AsyncListDifferWithPayload(@NonNull ListUpdateCallback listUpdateCallback,
+      @NonNull DiffUtil.ItemCallback<T> diffCallback) {
+    updateCallback = listUpdateCallback;
+    this.diffCallback = diffCallback;
+  }
+
+  @Nullable
+  private List<T> list;
+
+  /**
+   * Non-null, unmodifiable version of list.
+   * <p>
+   * Collections.emptyList when list is null, wrapped by Collections.unmodifiableList otherwise
+   */
+  @NonNull
+  private List<T> readOnlyList = Collections.emptyList();
+
+  // Max generation of currently scheduled runnable
+  private int maxScheduledGeneration;
+  private int maxFinishedGeneration;
+
+  /**
+   * Get the current List - any diffing to present this list has already been computed and
+   * dispatched via the ListUpdateCallback.
+   * <p>
+   * If a <code>null</code> List, or no List has been submitted, an empty list will be returned.
+   * <p>
+   * The returned list may not be mutated - mutations to content must be done through
+   * {@link #submitList(List)}.
+   *
+   * @return current List.
+   */
+  @NonNull
+  public List<T> getCurrentList() {
+    return readOnlyList;
+  }
+
+  /**
+   * Prevents any ongoing diff from dispatching results. Returns true if there was an ongoing
+   * diff to cancel, false otherwise.
+   */
+  public boolean cancelDiff() {
+    boolean diffInProgress = maxScheduledGeneration > maxFinishedGeneration;
+    maxFinishedGeneration = maxScheduledGeneration;
+    return diffInProgress;
+  }
+
+  /**
+   * Pass a new List to the AdapterHelper. Adapter updates will be computed on a background
+   * thread.
+   * <p>
+   * If a List is already present, a diff will be computed asynchronously on a background thread.
+   * When the diff is computed, it will be applied (dispatched to the {@link ListUpdateCallback}),
+   * and the new List will be swapped in.
+   *
+   * @param newList The new List.
+   */
+  @SuppressWarnings("WeakerAccess")
+  public void submitList(final List<T> newList) {
+    if (newList == list) {
+      // nothing to do
+      return;
+    }
+
+    // incrementing generation means any currently-running diffs are discarded when they finish
+    final int runGeneration = ++maxScheduledGeneration;
+
+    if (newList == null) {
+      //noinspection ConstantConditions
+      updateCallback.onRemoved(0, list.size());
+      list = null;
+      readOnlyList = Collections.emptyList();
+      return;
+    }
+
+    if (list == null) {
+      // fast simple first insert
+      updateCallback.onInserted(0, newList.size());
+      list = newList;
+      readOnlyList = Collections.unmodifiableList(newList);
+      return;
+    }
+
+    final DiffCallback<T> wrappedCallback = new DiffCallback<>(list, newList, diffCallback);
+
+    DIFF_EXECUTOR.execute(new Runnable() {
```

Comment:
This would be easier with coroutines :)

Reply:
yeah :( not ready to add that dependency yet though